### PR TITLE
[Driver][SYCL] refactor -fsycl-device-only behaviors

### PR DESCRIPTION
**Pull request by mdtoguchi**
Update -fsycl-device-only to go through the offload path, effectively
removing the host dependencies when it is used.  This allows for a
cleaner representation of SYCL device specific compilation needs.

### CODE REVIEW

File: clang/lib/Driver/Driver.cpp [hunk 1]
```diff
@@ -3700,10 +3708,25 @@ class OffloadingActionBuilder final {
       // The host depends on the generated integrated header from the device
       // compilation.
       if (CurPhase == phases::Compile) {
+        bool SYCLDeviceOnly = Args.hasArg(options::OPT_fsycl_device_only);
         for (Action *&A : SYCLDeviceActions) {
           DeviceCompilerInput =
               C.MakeAction<CompileJobAction>(A, types::TY_SYCL_Header);
-          A = C.MakeAction<CompileJobAction>(A, types::TY_LLVM_BC);
+          types::ID OutputType = types::TY_LLVM_BC;
+          if (SYCLDeviceOnly) {
```
Comment (AGindinson) [inline clang/lib/Driver/Driver.cpp:3716]: Doesn't this block make [this one](https://github.com/intel/llvm/blob/sycl/clang/lib/Driver/Driver.cpp#L5414) obsolete? If not, do you think there's a way to remove this duplication by some additional refactoring?

**Review by AGindinson — COMMENTED** [review]
A great improvement overall! Just a few questions on my side to fully understand the changes.

File: clang/lib/Driver/Driver.cpp [hunk 2]
```diff
@@ -792,9 +792,10 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
   // the -fsycl-targets, -fsycl-add-targets or -fsycl-link-targets option.
   // If -fsycl is supplied without any of these we will assume SPIR-V.
   // Use of -fsycl-device-only overrides -fsycl.
-  bool HasValidSYCLRuntime = (C.getInputArgs().hasFlag(options::OPT_fsycl,
-      options::OPT_fno_sycl, false) &&
-      !C.getInputArgs().hasArg(options::OPT_fsycl_device_only));
+  bool HasValidSYCLRuntime =
+      (C.getInputArgs().hasFlag(options::OPT_fsycl, options::OPT_fno_sycl,
+                                false) ||
+       C.getInputArgs().hasArg(options::OPT_fsycl_device_only));
```
[the remark this answers is not in the document]
Comment (mdtoguchi) [reply clang/lib/Driver/Driver.cpp:798]: From a usage standpoint, I don't feel there should be a requirement to use `-fsycl` with `-fsycl-device-only`.  When using `-fsycl` with `-fsycl-device-only`, `-fsycl-device-only` always takes precedence.  I've always seen `-fsycl` as an 'offload enabling' option, while `-fsycl-device-only` also enables offloading, the behavior is that of a single top level compilation.

File: clang/lib/Driver/Driver.cpp (hunk 1 above)
Comment (mdtoguchi) [reply clang/lib/Driver/Driver.cpp:3716]: It does make that block obsolete - I even removed it in this review, so I'm a little confused... :)

File: clang/lib/Driver/Driver.cpp [hunk 3]
```diff
@@ -3700,10 +3708,25 @@ class OffloadingActionBuilder final {
...
-          A = C.MakeAction<CompileJobAction>(A, types::TY_LLVM_BC);
+          types::ID OutputType = types::TY_LLVM_BC;
+          if (SYCLDeviceOnly) {
+            if (Args.hasArg(options::OPT_S)) {
+              OutputType = types::TY_LLVM_IR;
+            }
+            if (Args.hasFlag(options::OPT_fno_sycl_use_bitcode,
+                             options::OPT_fsycl_use_bitcode, false)) {
+              auto *BackendAction =
+                  C.MakeAction<BackendJobAction>(A, types::TY_LLVM_BC);
```
[the remark this answers is not in the document]
Comment (mdtoguchi) [reply clang/lib/Driver/Driver.cpp:3723]: I checked the behavior differences, and the results are identical.  I will update to use `CompileJobAction` for consistency.

File: clang/lib/Driver/Driver.cpp (hunk 1 above)
Comment (AGindinson) [reply clang/lib/Driver/Driver.cpp:3716]: Sorry for the miss :D At least we've gotten a truly independent double-check!

File: clang/lib/Driver/Driver.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (AGindinson) [reply clang/lib/Driver/Driver.cpp:798]: Sounds reasonable to me, thanks for elaborating on this.

**Review by AGindinson — APPROVED** [review]
LGTM!
As for the 1-line change in `libclc`, I get it @bader will be offline for the duration of the week. @romanovvlad, do you think this PR can be merged earlier as an exception?

**Pull request merged**